daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala

```diff
@@ -722,17 +722,15 @@ private[lf] final class Compiler(
       // This relies on interfaces having the same representation as the underlying template
       t.InterfacePrecondDefRef(impl._1)(env2.toSEVar(tmplArgPos))
     )
-    // TODO Clean this up as part of changing how we evaluate these
-    // https://github.com/digital-asset/daml/issues/11762
+
     val precondsArray: ImmArray[s.SExpr] =
-      (Iterator(translateExp(env2, tmpl.precond)) ++ implementsPrecondsIterator ++ Iterator(
-        s.SEValue.EmptyList
-      )).to(ImmArray)
-    val preconds = s.SEApp(s.SEBuiltin(SBConsMany(precondsArray.length - 1)), precondsArray.toList)
-    // We check precondition in a separated builtin to prevent
-    // further evaluation of agreement, signatories, observers and key
-    // in case of failed precondition.
-    let(env2, SBCheckPrecond(tmplId)(env2.toSEVar(tmplArgPos), preconds)) { (_, env) =>
+      (Iterator(translateExp(env2, tmpl.precond)) ++ implementsPrecondsIterator).to(ImmArray)
+
+    val preconds = precondsArray.foldLeft[s.SExpr](s.SEValue(SUnit))((acc, precond) =>
```

Comment:
(nit picking) 
```suggestion
    val preconds = precondsArray.foldLeft[s.SExpr](s.SEValue.SUnit)((acc, precond) =>
```
avoids an allocation.

Reply:
I am fine with the fold. 
Just wondering out loud if using nested `let` is not better than nested `application`. 
@nickchapman-da ?

Reply:
yep, fixed it.